### PR TITLE
WIP: Smarter cdn #1676

### PR DESCRIPTION
**Pull request by ralsina**
Try to be smarter and automatically use a CDN (or not) without having to do complex code in the templates.

### CODE REVIEW

**Comment by ralsina** [comment]
Having https://github.com/getnikola/nikola/pull/1680 would make this possible by making sure bundles are built after ``copy_assets`` and before ``pages``

**Comment by Kwpolska** [comment]
We could easily avoid dependency injection by using `cache/cdnjs` and `cache/cdncss` files that all pages would depend on.  The files would be created by the `bundles` plugin (specified as `targets`), and dependencies of all page builds.  (we could probably pass the lists from bundles to pages without reading the actual files on page generation)

File: scripts/get-cdnjs-list.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env python
+
+import json
+import subprocess
+
+links = {}
+data = subprocess.check_output(['links', '-dump', 'https://cdnjs.com/'])
```
Comment (Kwpolska) [inline scripts/get-cdnjs-list.py]: We do not have to scrape the website.  There is a nice .json file we could use with a little parsing (might need an input format change to account for different files and versions): https://github.com/cdnjs/cdnjs/wiki/Extensions%2C-Plugins%2C-Resources#developers

**Comment by ralsina** [comment]
I'd love @Aeyoun and @Kwpolska 's input on this one :-)

Currently only the bootstrap3 theme is adapted, but you can see how it gets simpler and more robust. Currently the mako template is not jinjify-able tho.

**Comment by Kwpolska** [comment]
> Currently the mako template is not jinjify-able tho.

This is probably colorbox translations’ fault.  I suggest using string formatting, like this (untested):

```python
cdn_url('jquery.colorbox::i18n/jquery.colorbox-{lang}.js',
        use_cdn,
        '/assets/js/colorbox-i18n/jquery.colorbox-{lang}.js',
        lang=colorbox_locales[lang])
```

(using `**kwargs` and `str.format` on the Python side)

File: scripts/get-cdnjs-list.py (hunk 1 above)
Comment (Aeyoun) [inline scripts/get-cdnjs-list.py]: There should be some config with an array of which libraries to include. Hard-coding a few into the default themes is not really flexible.

Comment (ralsina) [inline scripts/get-cdnjs-list.py]: @Aeyoun I don't understand. The list of available libraries cdnjs supports is quite extensive.

Comment (Aeyoun) [inline scripts/get-cdnjs-list.py]: @ralsina, but there is no setting to tell Nikola what libraries to load. Each library is hardcoded in the theme and not loaded based on a configurable setting.

Comment (ralsina) [inline scripts/get-cdnjs-list.py]: @Aeyoun I always assumed the theme author chooses what he loads...

**Comment by ralsina** [comment]
This approach is not convincing to me anymore.

I'll explore alternatives.

**Pull request closed without merging**